classes/village/VillageUpgradeManager.php

```diff
@@ -681,4 +715,26 @@ public static function boostResearch(System $system, Village $village, string $u
         $system->db->query("UPDATE `village_upgrades` SET `research_boosted` = 1, `research_progress_last_updated` = " . time() . ", `research_progress_required` = " . $village->upgrades[$upgrade_key]->research_progress_required . " WHERE `id` = {$village->upgrades[$upgrade_key]->id}");
         return "Research boosted for " . VillageUpgradeConfig::UPGRADE_NAMES[$upgrade_key] . "!";
     }
+
+    public static function calcResearchScore(System $system, Village $village): int {
```

Comment:
These are a count of total days completed construction/research time, currently just for the catch-up bonus with Ashes policy but also a useful way for us to glance at the DB and see how much progress villages are making.